src/java/org/apache/cassandra/io/sstable/SSTableSimpleWriter.java

```diff
@@ -36,64 +42,132 @@
-        assert key != null;
+        Preconditions.checkArgument(key != null, "Partition update cannot have null key");
 
-        // If that's not the current key, write the current one if necessary and create a new
-        // update for the new key.
-        if (!key.equals(currentKey))
+        // update for the first partition or a new partition
+        if (update == null || !key.equals(currentKey))
         {
+            // write the previous update if not absent
             if (update != null)
-                writePartition(update.build());
+                writePartition(update.build()); // might switch to a new sstable writer and reset currentSize
+
+            // and create a new update for the new key
+            currentSize += PartitionUpdate.serializer.serializedSize(createUpdateBuilder(key).build(),
+                                                                     format.getLatestVersion().correspondingMessagingVersion());
             currentKey = key;
-            update = new PartitionUpdate.Builder(metadata.get(), currentKey, columns, 4);
+            update = createSizeCountingPartitionUpdateBuilder(key, this::countRow, () -> {});
         }
 
-        assert update != null;
+        Preconditions.checkState(update != null, "Partition update to write cannot be null");
         return update;
     }
 
+    @Override
     public void close()
+    {
+        writeLastPartitionUpdate(update);
+        closeWriterIfPresent(writer);
+    }
+
+    private boolean shouldSwitchToNewWriter()
+    {
+        return writer == null || (maxSSTableSize > 0 && currentSize > maxSSTableSize);
+    }
+
+    private SSTableTxnWriter getOrSwitchWriter() throws IOException
+    {
+        if (shouldSwitchToNewWriter())
+        {
+            closeWriterIfPresent(writer);
+            writer = createWriter(null);
+            currentSize = 0;
+        }
+
+        return writer;
+    }
+
+    private void countRow(Row row)
+    {
+        // Note that the accounting of a row is a bit inaccurate (it doesn't take some of the file format optimization into account).
+        // In particular, what we count is closer to the serialized value, but it's debatable that it's the right thing
+        // to count since it will take a lot more space in memory.
+        currentSize += UnfilteredSerializer.serializer.serializedSize(row, helper, 0,
+                                                                      format.getLatestVersion().correspondingMessagingVersion());
+    }
+
+    private void writeLastPartitionUpdate(PartitionUpdate.Builder update)
     {
         try
         {
             if (update != null)
                 writePartition(update.build());
+        }
+        catch (Throwable t)
+        {
+            Throwable e = writer == null ? t : writer.abort(t);
+            Throwables.throwIfUnchecked(e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void closeWriterIfPresent(SSTableTxnWriter writer)
```

Comment:
NIT for a more succinct method name
```suggestion
    private void maybeCloseWriter(SSTableTxnWriter writer)
```